### ISSUE TITLE
java FCM certificate

### ISSUE THREAD

**Issue by culebras** [bytefish/FcmJava#16]
Hi,

This is not really a real issue, is more like a question related to FCM and Java.
When trying to send push notification to FCM platform from my Java FCM server app, I was facing issues with certificates not trusted.

I finally solved it downloading the certificate of FCM from googleapis.com and importing it into my JDK keystore file.

Did you face this kind of issues?

**Comment by culebras** [comment]
More data about the exception:

```
java.lang.RuntimeException: javax.net.ssl.SSLHandshakeException: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target

	at de.bytefish.fcmjava.client.http.HttpClient.post(HttpClient.java:51)
	at de.bytefish.fcmjava.client.FcmClient.post(FcmClient.java:99)
	at de.bytefish.fcmjava.client.FcmClient.send(FcmClient.java:80)

Caused by: javax.net.ssl.SSLHandshakeException: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:192)
	at sun.security.ssl.SSLSocketImpl.fatal(SSLSocketImpl.java:1949)
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:302)
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:296)
	at sun.security.ssl.ClientHandshaker.serverCertificate(ClientHandshaker.java:1509)
	at sun.security.ssl.ClientHandshaker.processMessage(ClientHandshaker.java:216)
	at sun.security.ssl.Handshaker.processLoop(Handshaker.java:979)
	at sun.security.ssl.Handshaker.process_record(Handshaker.java:914)
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:1062)
	at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1375)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1403)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1387)
	at org.apache.http.conn.ssl.SSLConnectionSocketFactory.createLayeredSocket(SSLConnectionSocketFactory.java:394)
	at org.apache.http.conn.ssl.SSLConnectionSocketFactory.connectSocket(SSLConnectionSocketFactory.java:353)
	at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:134)
	at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:353)
	at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:380)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:236)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184)
	at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:107)
	at de.bytefish.fcmjava.client.http.HttpClient.internalPost(HttpClient.java:103)
	at de.bytefish.fcmjava.client.http.HttpClient.post(HttpClient.java:49)
	... 33 more
Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:387)
	at sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:292)
	at sun.security.validator.Validator.validate(Validator.java:260)
	at sun.security.ssl.X509TrustManagerImpl.validate(X509TrustManagerImpl.java:324)
	at sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:229)
	at sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:124)
	at sun.security.ssl.ClientHandshaker.serverCertificate(ClientHandshaker.java:1491)
	... 54 more
Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.provider.certpath.SunCertPathBuilder.build(SunCertPathBuilder.java:141)
	at sun.security.provider.certpath.SunCertPathBuilder.engineBuild(SunCertPathBuilder.java:126)
	at java.security.cert.CertPathBuilder.build(CertPathBuilder.java:280)
	at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:382)
	... 60 more
```

**Comment by bytefish** [comment]
I would download the certificate first:

```sh
openssl s_client -showcerts -connect fcm.googleapis.com:443 < /dev/null | openssl x509 -outform PEM > cert.pem
```

And then import it into the Java Keystore:

```sh
keytool -import -alias googleapis -file cert.pem -keystore mycerts -storepass changeit
```

Does this work for you?

**Comment by bytefish** [comment]
If it doesn't help, could you please enable the SSL Debugging mode with the following JVM parameter and tell me the output:

```
 -Djavax.net.debug=ssl:handshake
```

**Comment by culebras** [comment]
Hi, actually I already tried that solution. I downloaded the certificate and import it into my keystore. The exception I attached was related to before doing that (I created the posts in wrong order).

The thing is that finally I found that it is working correctly without importing any extra certificate, using directly the JDK keystore. It was failing for me because I forgot that we were using different keystore for our project.

So in general, it will work correctly without adding any extra certificate, it is enough to use default JDK keystore (**cacerts** file, located usually in jdk_home/jre/lib/security).

I think we can close this issue. Thanks.
